### ISSUE TITLE
Interactions between Process exits, queries, piped io?

### ISSUE THREAD

**Issue by dubiousjim** [denoland/deno#3743]
What are the interactions between these methods:

```
Process#close()
Process#kill(signal)
Process#status()
```

The existing docs or a quick read of the source say that:

1. calling `status()` after `Process#kill(signal)` should still be ok (and is).
2. calling `status()` after `Process#close()` will fail with `BadResource: bad resource id`. So too will calling `close()` after `close()`.
3. Because of 1 and 2, presumably `Process#kill` doesn't (and shouldn't) implicitly invoke `Process#close`.
4. What if `Process#close` is invoked, and the process is still running? Would we expect this to `kill` the Process, or not? Currently it doesn't seem to do so, and perhaps that's the right design. Just wanted to make the behavior explicit.
5. Currently nothing stops us from `Process#kill`ing a Process that has already exited. If you're lucky, your OS will complain. If you're unlucky, the `pid` may have been recycled, and you'll be signaling some unknown new process. Wouldn't it be nicer for Deno to verify that if you call `Process#kill`, the Process hasn't already exited (either naturally, or because you or someone else killed it)?

	One way to do that would be to run the `sendAsync(dispatch.OP_RUN_STATUS, {rid})` on every Process right after it's started, and have the resulting Promise set a flag, which blocks `Process#kill`. If that seems too resource-intensive to do across the board, perhaps there's a way to make sure it's at least been done first, in those cases the user requests to `kill` the Process (or queries its status)?

Next, what are the interactions of the above methods with these:

```
Process#output()
Process#stderrOutput()
Process#stdin|stdout|stderr.close()
```

6. The docs say that calling `output()` and `stderrOutput()` schedule a `stdout.close()` and `stderr.close()`, respectively, after the pipes have been read.

7. What if `Process#close` is invoked, the Process has piped stdio, but `output()` and `stderrOutput()` haven't been invoked, and perhaps never will be? Do the stdio pipes get closed? If so, when: only when they're garbage collected?

8. In general, what is appropriate clean-up for a Process? Should one always explicitly `close()` it? (I expect it will be `closed()` automatically when GC'd; though that doesn't mean we shouldn't close it ourselves when we know we're done with it.) If it has open stdio pipes that we haven't read --- perhaps because we killed the Process --- do we need to explicitly `close()` them before closing the Process?

**Comment by dubiousjim** [comment]
What I'm trying to do is to make a specialized version of `run` that permits imposing a timeout, and also a max size on the piped stdout. (Corresponding to Node's child_process.execFile.) For the former, I need a better understanding of the coupling between these methods --- and probably some more hooks into the implementation, too. (For the latter, I need a version of Deno.readAll that takes a maxBytes argument.)

**Comment by dubiousjim** [comment]
I've experimented a lot with the current implementation of Deno.run, and found it spits out errors / fails silently depending on when the process and/or its stdout are closed. I worked up a more robust implementation, which can be found [at this gist](https://gist.github.com/dubiousjim/f0d8c2299c5ea7f94e249553c6335323). It's in JS not TypeScript, and it needs some cleaning up and tests. (But I've tested it extensively here.)

One main perk is it's more robust about settling the promises from Process.status and Process.output (and .stderrOutput), no matter when the process and/or the stdout is closed.

A second perk is it has an optional maxBuffer limit on the stdout.

I intend to work a timeout mechanism in there too. I think the needed hooks are already there. Will get to that later.

The reason I'm posting this here now is that we're getting close to the Deno 1.0 release, and at that point I expect the API will become less fluid. When writing this version of run/Process, I had to re-work the Deno.Buffer.readFrom and Deno.readAll interfaces. It'd be nice if those changes could land in Deno, so that I didn't have to duplicate them. If so, those would be breaking changes. It would be great if someone could have a look at this and consider altering the API in these ways now --- before we hit the 1.0 slowdown.

I'll also work on polishing this work on run/Process and getting it into a shape to be a proper pull request. But for the moment, I just wanted to say something about the API. In the current gist, I have two functions Buffer.readFrom and Buffer.readFromN (taking a max nbytes to read), but those might reasonably be merged into a single function taking an optional parameter. I also have two functions readAll and readAllN (also taking a max nbytes to read), but these also might reasonably be merged.

Adding optional parameters to the Buffer.readFrom and readAll with enhanced behavior isn't a breaking change. What is breaking in my implementations is this:

* readAll no longer returns a `Promise<Uint8Array>`. Instead it returns a promise of an object of the form `{closed?:boolean, aborted?:any, contents:Uint8Array }`. The `closed` field is present whenever the Buffer.readFrom promise resolved successfully, and is set to `true` in case that caught a `BadResource` error while trying to read (else it's set to `false`). The `aborted` field is present when the Buffer.readFrom promise was rejected, and is set to the `reason` value given when rejecting. The `contents` field is always present, as the Buffer may have read some valid data before the readFrom promise was rejected.

* Buffer.readFrom needs to signal in some way whether it's returning from its `catch` branch or its main loop. I did this by returning a negative value when it returns from the `catch` branch. We take the two's complement of the number that we'd normally return, because the normal number might be 0. We can't implement the two's complement as just `~n` though, because `n` might be (almost) any 32 bit int. So we implement it as `-n-1`.

As I said, I'd be happy to discuss the other changes and work towards getting them into a form that can be pulled into this project. But I wanted to propose these API changes asap.
